addons/website_slides/static/src/js/slides_course_quiz.js

```diff
@@ -73,33 +78,28 @@ odoo.define('website_slides.quiz', function (require) {
                 self._renderAnswers();
                 self._renderAnswersHighlighting();
                 self._renderValidationInfo();
-                new CourseJoinWidget(self, self.channel.channelId).appendTo(self.$('.o_wslides_course_join_widget'));
+                var courseJoinWidget = new CourseJoinWidget(self, self.channel.channelId,true);
+                return courseJoinWidget;
+            }).then(function(courseJoinWidget){
+                if(self.notMember && !self.publicUser){
```

Comment:
Why is this in a `then` clause?

Looks like this code could be right after `var courseJoinWidget = new CourseJoinWidget(self, self.channel.channelId,true);`

Reply:
Because I wanted to make sur the widget was created before using .appendTo and .joinCourse. As I Understood JS is synchronous and you need to force asynchronous behavior with promises. Isn't this how it has to be done?

Reply:
JS is asynchronous and you need to force synchronous behavior with promises*

Yes, it's indeed the case, but you don't need it here.

When you instantiate a widget, it calls the "init" method of that widget synchronously.
Then, when you use your `appendTo` just below, it calls the "start" method of the widget, that one is async.
So you can call appendTo right after instantiating the widget without any issue.

What you're doing correctly is putting all that in the  "then" phase of your parent widget, meaning it's DOM is ready.

Ideally, that whole block of code should be extracted in a sub-method, just like the `_renderX` above.
So you would have a `_renderJoinWidget()`.

If that's not clear, don't hesitate to ask :)

Reply:
I think it's clear thanks, I created the _renderJoinWidget() methode.